bundles/org.openhab.binding.bmwconnecteddrive/src/main/java/org/openhab/binding/bmwconnecteddrive/internal/handler/ConnectedDriveProxy.java

```diff
@@ -236,89 +256,203 @@ RemoteServiceHandler getRemoteServiceHandler(VehicleHandler vehicleHandler) {
+            HttpURLConnection.setFollowRedirects(false);
+            HttpURLConnection con = (HttpURLConnection) url.openConnection();
+            con.setRequestMethod("POST");
+            con.setRequestProperty(HttpHeader.CONTENT_TYPE.toString(), CONTENT_TYPE_URL_ENCODED);
+            con.setRequestProperty(HttpHeader.CONNECTION.toString(), KEEP_ALIVE);
+            con.setRequestProperty(HttpHeader.HOST.toString(),
+                    BimmerConstants.API_SERVER_MAP.get(configuration.region));
+            con.setRequestProperty(HttpHeader.AUTHORIZATION.toString(),
+                    BimmerConstants.LEGACY_AUTHORIZATION_VALUE_MAP.get(configuration.region));
+            con.setRequestProperty(CREDENTIALS, BimmerConstants.LEGACY_CREDENTIAL_VALUES);
+            con.setDoOutput(true);
+
+            OutputStream os = con.getOutputStream();
+            byte[] input = getAuthEncodedData().getBytes("utf-8");
+            os.write(input, 0, input.length);
+
+            BufferedReader br = new BufferedReader(new InputStreamReader(con.getInputStream(), "utf-8"));
+            StringBuilder response = new StringBuilder();
+            String responseLine = null;
+            while ((responseLine = br.readLine()) != null) {
+                response.append(responseLine.trim());
+            }
+            token.setMyBmwApiUsage(false);
+            return tokenFromUrl(con.getHeaderField(HttpHeader.LOCATION.toString()));
+        } catch (IOException e) {
+            logger.warn("{}", e.getMessage());
+        }
+        return false;
     }
 
-    void tokenFromUrl(String encodedUrl) {
+    public boolean tokenFromUrl(String encodedUrl) {
+        final StringBuilder result = new StringBuilder();
         final MultiMap<String> tokenMap = new MultiMap<String>();
         UrlEncoded.decodeTo(encodedUrl, tokenMap, StandardCharsets.US_ASCII);
         tokenMap.forEach((key, value) -> {
             if (value.size() > 0) {
                 String val = value.get(0);
                 if (key.endsWith(ACCESS_TOKEN)) {
                     token.setToken(val.toString());
+                    result.append(true);
                 } else if (key.equals(EXPIRES_IN)) {
                     token.setExpiration(Integer.parseInt(val.toString()));
                 } else if (key.equals(TOKEN_TYPE)) {
                     token.setType(val.toString());
                 }
             }
         });
+        return Boolean.valueOf(result.toString());
+    }
+
+    public static String codeFromUrl(String encodedUrl) {
+        final MultiMap<String> tokenMap = new MultiMap<String>();
+        UrlEncoded.decodeTo(encodedUrl, tokenMap, StandardCharsets.US_ASCII);
+        final StringBuilder codeFound = new StringBuilder();
+        tokenMap.forEach((key, value) -> {
+            if (value.size() > 0) {
+                String val = value.get(0);
+                if (key.endsWith(CODE)) {
+                    codeFound.append(val.toString());
```

Comment:
Can we find multiple or should we just return the first one? Maybe you should use streams instead of just a for-each. 

Something like (pseudo-code):

```
tokenMap.stream()
   .filter( key.endsWith(CODE))
   .map ( val.toString)
   .firstOrDefault("")

Reply:
There's no stream() function available for MultiMap which is filled by decodeTo
```
        final MultiMap<String> tokenMap = new MultiMap<String>();
        UrlEncoded.decodeTo(encodedUrl, tokenMap, StandardCharsets.US_ASCII);
```